django/db/models/functions.py

```diff
@@ -83,6 +83,28 @@ def _paired(self, expressions):
         return ConcatPair(expressions[0], self._paired(expressions[1:]))
 
 
+class Greatest(Func):
+    """
+    Chooses the maximum non-null expression and returns it.
+    """
+    function = 'GREATEST'
+
+    def __init__(self, *expressions, **extra):
+        if len(expressions) < 1:
+            raise ValueError('Greatest must take at least one expression')
+        super(Greatest, self).__init__(*expressions, **extra)
+
+    def as_sqlite(self, compiler, connection):
+        self.coalesce()
+        return super(Greatest, self).as_sql(compiler, connection, function='MAX')
+
+    def coalesce(self):
+        # any null results in null for the whole expression, wrap with coalesce
+        expressions = [
+            Coalesce(expression, Value('')) for expression in self.get_source_expressions()]
+        self.set_source_expressions(expressions)
```

Comment:
The problem here is that you can't just use `Value('')` for the default. If you're doing `GREATEST(date_field, other_date_field)` then coalescing a date type with a char type is going to produce an error.

The type itself will probably have to accept a default.

```
sentinel = object()
def __init__(self, *expressions, **kwargs):
    ifnull = kwargs.pop('ifnull', sentinel)
    if ifnull == sentinel:
        raise ValueError('ifnull is required')
    if ifnull is None:
        # user has asked NOT to use coalesce
    else:
        self.ifnull = self._parse_expression(ifnull)
```

And then you would use `Coalesce(expression, self.ifnull)` in the coalesce method, or completely skip calling the coalesce method if `ifnull is None`.

This is just one idea, but probably the best one I have right now. I don't really like forcing a user to provide an `ifnull` though, because it feels like we're disadvantaging the user.

Another idea would be to use a backend feature. Something like `greatest_least_uses_nulls`, and then the tests could switch on that feature flag to provide different test results.

I'd probably like to get a rough consensus on which way to go here.